lib/evmone/eof.cpp

```diff
@@ -90,37 +150,270 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+            {
+                auto jumpdest_opt =
+                    check_rjump_destination(&code[i + 1 + 1 + static_cast<uint16_t>(k)], post_pos);
+
+                if (jumpdest_opt)
+                    rjumpdests.push_back(jumpdest_opt.value());
+                else
+                    return false;
+            }
+        }
+
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+/// Requires that the input is validated against truncation.
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOFCodeType>& code_types)
+{
+    assert(!code.empty());
+
+    // Special values used for detecting errors.
+    constexpr int32_t LOC_UNVISITED = -1;
+    constexpr int32_t LOC_IMMEDIATE = -2;
+
+    // Stack height in the header is limited to uint16_t, but keeping larger size for ease of
+    // calculation.
+    std::vector<int32_t> stack_heights = std::vector<int32_t>(code.size(), LOC_UNVISITED);
+    stack_heights[0] = code_types[func_index].inputs;
+
+    std::stack<size_t> worklist;
+    worklist.push(0);
+
+    while (!worklist.empty())
+    {
+        const auto i = worklist.top();
+        worklist.pop();
+
+        const auto opcode = static_cast<Opcode>(code[i]);
+
+        auto stack_height_required = instr::traits[opcode].stack_height_required;
+        auto stack_height_change = instr::traits[opcode].stack_height_change;
+
+        if (opcode == OP_CALLF)
+        {
+            const auto fid = read_uint16_be(&code[i + 1]);
+
+            if (fid >= code_types.size())
+                return {EOFValidationError::invalid_code_section_index, 0};
+
+            stack_height_required = static_cast<int8_t>(code_types[fid].inputs);
+            stack_height_change =
+                static_cast<int8_t>(code_types[fid].outputs - stack_height_required);
+        }
+
+        auto stack_height = stack_heights[i];
+        assert(stack_height != LOC_UNVISITED);
+
+        if (stack_height < stack_height_required)
+            return {EOFValidationError::stack_underflow, -1};
+
+        // Target locations of control flow.
+        std::vector<size_t> successors;
+
+        // Determine size of immediate, including the special case of RJUMPV.
+        const size_t imm_size = (opcode == OP_RJUMPV) ? (1 + /*count*/ size_t{code[i + 1]} * 2) :
+                                                        instr::traits[opcode].immediate_size;
+
+        // Mark immediate locations.
+        const auto beg = stack_heights.begin() + static_cast<int32_t>(i) + 1;
+        std::fill_n(beg, imm_size, LOC_IMMEDIATE);
+
+        // Check validity of next instruction. We skip RJUMP and terminating instructions.
```

Comment:
Entire successor finding is rather independent, I would extract it to a separate function to make `validate_max_stack_height` simpler.